### PR TITLE
Parallel jobs on HPC cluster: `srun` versus `mpirun` submission commands

### LINKED ISSUE #2203: Running FV3GFS simulations in a segmeted mode
Machine learning (ML) workflow consists of several stages, some of which require running FV3GFS model. It is preferrable to run the model in a segmented mode due to the time limits on Stellar cluster. To accomplish this task I had to modify several lines in the run.py script 

https://github.com/ai2cm/fv3net/blob/3b1ff5c9378f6a70c8b760e4f02df373d32de780/workflows/prognostic_c48_run/runtime/segmented_run/run.py#L38-L46

as   '[
                    "srun",
                    "--export=ALL",
                    "--ntasks=" + str(nprocs),
                    "--cpus-per-task=1",
                    sys.executable,
                    "-m",
                    "mpi4py",
                    runfile.absolute().as_posix(),
                ]'

Even though this code modification allowed me to run the model, a more elegant and user-friendly solution is necessary to remain consistent with the overall structure of the code.

### PR DESCRIPTION
**Pull request by mrudko**
Performing parallel simulations with `FV3GFS` model rely on the availability of the MPI (Message Passing Interface) library. Scheduling and submitting jobs on HPC (High Performance Computing) cluster is accomplished via SLURM (Simple Linux Utility for Resource Management) job scheduler. In contrast to cloud platforms, clusters are configured to use the `srun` submission command, which is the SLURM analog to `mpirun`. This PR introduces the changes to the code to toggle between `mpirun` versus `sun` submission commands at user's disposal.

Added public API:
- Added option `wrapper` to both `append` and `run_native` commands in the command line interface of the `runfv3` tool (`cli.py` script). The option `wrapper` is needed to toggle between the submission commands on the HPC cluster versus Cloud.

Refactored public API:
- `append_segment_to_run_url` and `run_segment` accept a new argument `wrapper` (`append.py` and `run.py` scripts) 
- the variable `command` was introduced to the `run_segment` function (`run.py` script). 

- [x] Tests added
   A new test has been developed to validate the value of "mpi_launcher" within "compose_simulation_command" function.

Resolves #2203

### CODE REVIEW

File: workflows/prognostic_c48_run/runtime/segmented_run/append.py
Comment (mrudko) [reply workflows/prognostic_c48_run/runtime/segmented_run/append.py]: Thanks @spencerkclark for your valuable valuable comments. I just pushed another round of commits, and as of now, two tests have failed on CircleCI. One failure is due to linting errors - the "help messages" are way too long. Apologies for not issuing the command `make lint` before pushing the commits (I actually already addressed this issue). The root of the second failure is in passing extra argument to the function `append` in `cli.py` script. Can we talk offline about it?

Comment (spencerkclark) [reply workflows/prognostic_c48_run/runtime/segmented_run/append.py]: No worries regarding the linting failures.  

The errors in running the prognostic run image tests come from a subtle issue.  The clue is in this line from the errors:
```
{"message": "[mpiexec@5121e214c9f7] match_arg (utils/args/args.c:163): unrecognized argument n 6", "logging.googleapis.com/labels": {"kind": "fortran", "model_time": null}, "severity": "DEBUG", "model_time": null}
```
This indicates an issue with the command being submitted via the `subprocess` module (namely that the number of MPI ranks is not being specified properly).  Happy to talk about this more offline.

File: workflows/prognostic_c48_run/runtime/segmented_run/run.py
Comment (spencerkclark) [inline workflows/prognostic_c48_run/runtime/segmented_run/run.py:27]: This is the subtle change required for the tests to pass (this also makes things consistent with how they were written before for the `"mpirun"` case):
```suggestion
    command = ["-n", str(nprocs), sys.executable, "-m", "mpi4py", runfile.absolute().as_posix()]
```
See the note regarding "how to break a shell command into a sequence of arguments" in the [`subprocess.Popen` documentation](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) for more details.  The documentation admits this "may not be obvious!"

File: workflows/prognostic_c48_run/docs/local-execution.rst
Comment (spencerkclark) [inline workflows/prognostic_c48_run/docs/local-execution.rst:91]: I think we do not necessarily need to cover this case, since it is covered above (using the default value for the option).
```suggestion
```

File: workflows/prognostic_c48_run/docs/local-execution.rst
Comment (spencerkclark) [inline workflows/prognostic_c48_run/docs/local-execution.rst:98]: This will ensure the example is formatted as a code block in the documentation (see [this section of the Sphinx documentation on "Literal blocks"](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks)):
```suggestion
    An example of running model with 5 segments on an HPC cluster::

        for i in {1..5}
        do
           runfv3 append --mpi_launcher srun /absolute/file/path
        done
```

File: workflows/prognostic_c48_run/docs/local-execution.rst
Comment (spencerkclark) [inline workflows/prognostic_c48_run/docs/local-execution.rst:84]: I recommend putting `mpi_launcher` and `append` in backticks, so they will be formatted as code in the documentation (see the section on ["Inline markup" in the Sphinx documentation](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#inline-markup)).  I edited the text a bit too to be more specific about what this option does.
```suggestion
    An option ``mpi_launcher`` has been added to the subcommand ``append``. This
    option allows users to specify the system-specific MPI launcher used when running
    the model. By default it takes the value "mpirun", which is relevant for running
    in Docker. The only other currently supported value is "srun", which is relevant
    when running on SLURM-based High Performance Computing (HPC) systems. 
```

File: workflows/prognostic_c48_run/docs/local-execution.rst
Comment (spencerkclark) [inline workflows/prognostic_c48_run/docs/local-execution.rst:127]: A few minor formatting / spelling edits here as well:
```suggestion
   For the purposes of performing simulations on both Docker and HPC platforms,
   the subcommand ``run-native`` was supplemented with the ``mpi_launcher``
   option. Please see the example below on how to use it on an HPC cluster::

       runfv3 run-native fv3config.yaml path/to/local/rundir --mpi_launcher srun
```

File: workflows/prognostic_c48_run/runtime/segmented_run/run.py
Comment (spencerkclark) [inline workflows/prognostic_c48_run/runtime/segmented_run/run.py:36]: It would be good to stick with the literal name of the option here.  I think a [`ValueError`](https://docs.python.org/3/library/exceptions.html#ValueError) also makes the most sense to raise, rather than a generic `Exception`:
```suggestion
        raise ValueError("Unrecognized mpi_launcher option. Please choose between 'mpirun' or 'srun'.")
```

File: workflows/prognostic_c48_run/runtime/segmented_run/run.py
Comment (spencerkclark) [inline workflows/prognostic_c48_run/runtime/segmented_run/run.py:34]: Nice!  Using `insert` is a clean way of doing this.

File: workflows/prognostic_c48_run/runtime/segmented_run/run.py
Comment (spencerkclark) [inline workflows/prognostic_c48_run/runtime/segmented_run/run.py:30]: nit: since this is needed for both the supported `mpi_launcher` options, it might make sense to just put in the `command` at the start.

**Review by spencerkclark — COMMENTED** [review]
Thanks for adding those notes to the documentation!  I made a few suggestions there ([this "primer"](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#restructuredtext-primer) is a useful starting place for learning about the syntax used by Sphinx to format documentation).

I also made suggestion which I think should fix the failing tests.

**Comment by mrudko** [comment]
Apparently, Circle CI complains about the newly developed regression test for `compose_simulation_command` function. This goes at odds with the output of testing in my local home directory.

```
(fv3net) [mr7417@stellar-amd segmented_run]$ pytest -k "not gcs" test_append.py 
========================================================= test session starts ==========================================================
platform linux -- Python 3.8.10, pytest-4.6.11, py-1.11.0, pluggy-0.13.1
rootdir: /home/mr7417/ML_workflow/model_environment/fv3net, inifile: pytest.ini
plugins: hypothesis-6.17.4, regtest-1.4.4
collected 5 items / 1 deselected / 4 selected                                                                                          

test_append.py ....                                                                                                              [100%]

=========================================================== warnings summary ===========================================================
/home/mr7417/ML_workflow/model_environment/conda/envs/fv3net/lib/python3.8/site-packages/intake/utils.py:230
  /home/mr7417/ML_workflow/model_environment/conda/envs/fv3net/lib/python3.8/site-packages/intake/utils.py:230: DeprecationWarning: In a future release of intake, the intake.registry will not be directly mutable. Use intake.register_driver.
    warnings.warn(

/home/mr7417/ML_workflow/model_environment/conda/envs/fv3net/lib/python3.8/site-packages/intake/utils.py:230
  /home/mr7417/ML_workflow/model_environment/conda/envs/fv3net/lib/python3.8/site-packages/intake/utils.py:230: DeprecationWarning: In a future release of intake, the intake.container_map will not be directly mutable. Use intake.register_container.
    warnings.warn(

/home/mr7417/ML_workflow/model_environment/conda/envs/fv3net/lib/python3.8/site-packages/intake/source/discovery.py:58
  /home/mr7417/ML_workflow/model_environment/conda/envs/fv3net/lib/python3.8/site-packages/intake/source/discovery.py:58: PendingDeprecationWarning: The option `do_package_scan` may be removed in a future release.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========================================= 4 passed, 1 deselected, 3 warnings in 10.04 seconds ==========================================
```

File: workflows/prognostic_c48_run/runtime/segmented_run/cli.py
Comment (spencerkclark) [inline workflows/prognostic_c48_run/runtime/segmented_run/cli.py:58]: If a string needs to be split over multiple lines, I recommend using this syntax instead.  I think using `"\"` may inadvertently include the leading spaces at the start of the second line.
```suggestion
    help="MPI launcher used to run the simulation. "
         "Supported values are 'mpirun' and 'srun'.",
```
Since the same string is used farther down in the `run-native` command definition, you could consider defining it as a module-level constant.

File: workflows/prognostic_c48_run/runtime/segmented_run/cli.py
Comment (spencerkclark) [inline workflows/prognostic_c48_run/runtime/segmented_run/cli.py:71]:
```suggestion
    help="MPI launcher used to run the simulation. "
         "Supported values are 'mpirun' and 'srun'.",
```

**Review by spencerkclark — COMMENTED** [review]
Thanks for the updates @mrudko -- this is looking pretty close.

A regression-style test is maybe not the best choice in this case because `sys.executable` and `runfile.absolute().as_posix()` will differ depending on the platform (so capturing what the command looks like on one platform will not translate to what it looks like on another platform).  I recommend testing this by defining the expected outputs within the test itself (dynamically setting the path to the executable and the path to the run file).

Also as a reminder--though it is a bit moot, since I think a different style test should be used--when using a regression-style test please remember to commit the associated `*.out` reference files, which define the expected outputs of the tests.

**Review by spencerkclark — COMMENTED** [review]
Thanks @mrudko--almost there--could you add a test case for the error resulting from providing an unsupported `mpi_launcher` value too?

**Review by spencerkclark — COMMENTED** [review]
Thanks for fixing the test and adding a case for the error!  This looks good to me now.  There are just the linting errors to take care of:

```
black....................................................................Failed
- hook id: black
- files were modified by this hook

reformatted /home/circleci/project/workflows/prognostic_c48_run/runtime/segmented_run/cli.py
reformatted /home/circleci/project/workflows/prognostic_c48_run/tests/segmented_run/test_append.py
All done! ✨ 🍰 ✨
2 files reformatted, 700 files left unchanged.

Check for added large files..............................................Passed
Trim Trailing Whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing workflows/prognostic_c48_run/docs/local-execution.rst

flake8...................................................................Failed
- hook id: flake8
- exit code: 1

workflows/prognostic_c48_run/tests/segmented_run/test_append.py:71:23: E711 comparison to None should be 'if cond is None:'

flake8 __init__.py files.................................................Passed
mypy.....................................................................Passed
```

I think you'll need to fix the `flake8` issue manually (per my suggestion below), but the other formatting issues should be addressable by running `make lint` and committing the modified files.

**Review by spencerkclark — APPROVED** [review]
Thanks @mrudko!  This looks good to me now.  I'll try and get another pair of eyes on it today.

**Comment by spencerkclark** [comment]
@mrudko it should end up being OK one way or another (worst case we can always modify or revert the changes in a follow up PR if need be), but next time please double check with me before merging.  I did ask @oliverwm1 to take a look at this yesterday, but he did not have a chance to yet.  For this PR in particular I wanted him to take a look since we had not discussed this solution with him ahead of time.

I could be being a little overly cautious, but for these changes which touch fairly fundamental parts of our codebase I like to solicit feedback a little more widely.

**Comment by mrudko** [comment]
Hi Spencer,

Apologies for not letting you know. Since you approved all the code changes, I was assuming that that was my job to go ahead and merge it. I will keep you informed next time.

Mike

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Pull request merged**
